### ISSUE TITLE
How to use WebFlux for BatchLoader?  

### ISSUE THREAD

**Issue by FutureElement** [graphql-java/graphql-java#1039]
`@Component
public class GroupBatchLoader implements BatchLoader<String, Group> {

    private final KanbanService groupService;

    @Autowired
    public GroupBatchLoader(KanbanService groupService) {
        this.groupService = groupService;
    }

    @Override
    public CompletionStage<List<Group>> load(List<String> keys) {
        Flux<Group> flux = groupService.getGroupsByProjectIds(keys);
        return ???
    }
}`

groupService return the type of Flux<Group>, how to convert to  CompletionStage<List<Group>>?
or there is an other way to use webflux for batchLoader？

**Comment by bbakerman** [comment]
A flux is a reactive stream of results.  The list of Groups above can come down some time in the future, technically sporadically and not as a complete list as once.

A CompletionStage<List<Group>> is a promise to give a complete list of groups some time in the future.  Once the result does arrive, it must be there as a complete unit.

So reactive streams (such as Flux) and CompletionStages share some similarities but they differ in the way results turn up.

If you have a reactive backend of data, you MUST fully read that Flux (reactive stream) before you can consider the CompletionStage done.

graphql-java is "async" not fully reactive.

What I would do is create your own helper that "creates a uncompleted CompleteableFuture" and has code that that observes Flux until it signals it is done.  You buffer up the groups in this case and then publish the List<Groups> on the CompletableFuture.

In really rough psudeo code it might

```
CompleteableFuture cf = new  CompleteableFuture();
Flux<Group> flux = groupServer.get(ids)
    List<Group> buffer = new List()
flux.subscribe(new Subscription() {
   onObject(group) {
        buffer.add(group)
   }

   onEnd() {
        cf.complete(buffer);
   }    
}
return cf;
```

This is really rough code.  Off hand I dont recall how Flux is observed and projectreactor javadoc is down right now as I type.

Basically you "observer all the values in the Flux" and buffer then up.  When it finishes you can "complete" the CF with the buffered list

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#complete-T-

Remember also there can be exceptions and you need to complete the CF exceptionally

    cf.completeExceptionally(someException)

**Comment by FutureElement** [comment]
thanks very much

**Comment by FutureElement** [comment]
how about like this
`
@Override
public CompletionStage<List<Group>> load(List<String> keys) {
    Flux<Group> flux = groupService.getGroupsByProjectIds(keys);
    return flux.collectList().toFuture();
}
`
